{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pycytominer General Usage Walkthrough!\n",
    "\n",
    "Welcome to this walkthrough where we will guide you through the process of extracting single cell morphology features using the [`pycytominer`](https://github.com/cytomining/pycytominer) API.\n",
    "\n",
    "For this walkthrough, we will be working with the NF1-Schwann cell morphology dataset. \n",
    "If you would like the more information about this dataset, you can refer to this [repository](https://github.com/WayScience/Benchmarking_NF1_data)\n",
    "\n",
    "From the mentioned repo, we specifically used this [dataset](https://github.com/WayScience/Benchmarking_NF1_data/tree/main/4_processing_features/data/Plate2/CellProfiler) and the associated [metadata](https://github.com/WayScience/Benchmarking_NF1_data/tree/main/3_extracting_features/metadata) to generate the walkthrough. \n",
    "\n",
    "\n",
    "Let's get started with the walkthrough below!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# pycytominer imports\n",
    "from pycytominer.cyto_utils.cells import SingleCells\n",
    "from pycytominer import annotate, normalize, feature_select\n",
    "\n",
    "# ignore warnings\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## About the inputs\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this section, we will set up the expected input and output paths that will be generated throughout this walkthrough. Let's take a look at the explanation of these inputs and outputs.\n",
    "\n",
    "For this workflow, we have two main inputs:\n",
    "\n",
    "- **plate_data**: This contains the quantified single-cell morphology features that we'll be working with.\n",
    "- **plate_map**: This contains additional information related to the cells, providing valuable context of our single-cell morphology dataset.\n",
    "\n",
    "Now, let's explore the outputs generated in this workflow. In this walkthrough, we will be generating four profiles:\n",
    "\n",
    "- **sc_profile_path***: This refers to the single-cell morphology profile that will be generated.\n",
    "- **anno_profile_path**: This corresponds to the annotated single-cell morphology profile.\n",
    "- **norm_profile_path**: This represents the normalized single-cell morphology profile.\n",
    "- **feat_profile_path**: Lastly, this refers to the selected features from the single-cell morphology profile.\n",
    "\n",
    "These profiles will serve as important outputs that will help us analyze and interpret the single-cell morphology data effectively. Now that we have a clear understanding of the inputs and outputs, let's proceed further in our walkthrough."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setting file paths\n",
    "data_dir = pathlib.Path(\"./data/\").resolve(strict=True)\n",
    "metadata_dir = (data_dir / \"metadata\").resolve(strict=True)\n",
    "out_dir = pathlib.Path(\"results\")\n",
    "out_dir.mkdir(exist_ok=True)\n",
    "\n",
    "# input file paths\n",
    "plate_data = pathlib.Path(\"./data/nf1_data.sqlite\").resolve(strict=True)\n",
    "plate_map = (metadata_dir / \"platemap_NF1_CP.csv\").resolve(strict=True)\n",
    "\n",
    "# setting output paths\n",
    "sc_profile_path = out_dir / \"nf1_single_cell_profile.csv.gz\"\n",
    "anno_profile_path = out_dir / \"nf1_annotated_profile.csv.gz\"\n",
    "norm_profile_path = out_dir / \"nf1_noramlzied_profile.csv.gz\"\n",
    "feat_profile_path = out_dir / \"nf1_features_profile.csv.gz\"\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating Merged Single-cell Morphology Dataset\n",
    "\n",
    "In this section of the walkthrough, our goal is to load the NF1 dataset and create a merged single-cell morphology dataset.\n",
    "\n",
    "Currently, the NF1 dataset is stored in an `sqlite` format, where each table represents a different compartment, such as Image, Cell, Nucleus, and Cytoplasm.\n",
    "To achieve this, we will utilize the SingleCells class, which offers a range of functionalities specifically designed for single-cell analysis. You can find detailed documentation on these functionalities [here](https://pycytominer.readthedocs.io/en/latest/pycytominer.cyto_utils.html#pycytominer.cyto_utils.cells.SingleCells).\n",
    "\n",
    "However, for our purpose in this walkthrough, we will focus on using the SingleCells class to merge all the tables within the NF1 sqlite file into a merged single-cell morphology dataset."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Updating defaults\n",
    "Before we proceed further, it is important to update the default parameters in the SingleCells class to accommodate the table name changes in our NF1 dataset.\n",
    "\n",
    "Since the table names in our NF1 dataset differ from the default table names recognized by the SingleCells class, we need to make adjustments to ensure proper recognition of these table name changes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update compartment names and strata\n",
    "strata = [\"Image_Metadata_Well\", \"Image_Metadata_Plate\"]\n",
    "compartments = [\"Per_Cells\", \"Per_Cytoplasm\", \"Per_Nuclei\"]\n",
    "\n",
    "# Updating linking columns for merging all compartments\n",
    "linking_cols = {\n",
    "    \"Per_Cytoplasm\": {\n",
    "        \"Per_Cells\": \"Cytoplasm_Parent_Cells\",\n",
    "        \"Per_Nuclei\": \"Cytoplasm_Parent_Nuclei\",\n",
    "    },\n",
    "    \"Per_Cells\": {\"Per_Cytoplasm\": \"Cells_Number_Object_Number\"},\n",
    "    \"Per_Nuclei\": {\"Per_Cytoplasm\": \"Nuclei_Number_Object_Number\"},\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have stored the updated the parameters, we can use them as inputs for SingleCells class to proceed with the merging of all the NF1 sqlite tables into a single consolidated dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metadata_ImageNumber</th>\n",
       "      <th>Image_Metadata_Plate</th>\n",
       "      <th>Image_Metadata_Well</th>\n",
       "      <th>Cytoplasm_Number_Object_Number</th>\n",
       "      <th>Cytoplasm_AreaShape_Area</th>\n",
       "      <th>Cytoplasm_AreaShape_BoundingBoxArea</th>\n",
       "      <th>Cytoplasm_AreaShape_BoundingBoxMaximum_X</th>\n",
       "      <th>Cytoplasm_AreaShape_BoundingBoxMaximum_Y</th>\n",
       "      <th>Cytoplasm_AreaShape_BoundingBoxMinimum_X</th>\n",
       "      <th>Cytoplasm_AreaShape_BoundingBoxMinimum_Y</th>\n",
       "      <th>...</th>\n",
       "      <th>Nuclei_Texture_Variance_DAPI_3_02_256</th>\n",
       "      <th>Nuclei_Texture_Variance_DAPI_3_03_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_00_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_01_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_02_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_03_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_00_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_01_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_02_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_03_256</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>1</td>\n",
       "      <td>32065.0</td>\n",
       "      <td>95944.0</td>\n",
       "      <td>1018.0</td>\n",
       "      <td>382.0</td>\n",
       "      <td>750.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1393.410857</td>\n",
       "      <td>1331.583275</td>\n",
       "      <td>653.826838</td>\n",
       "      <td>618.063979</td>\n",
       "      <td>606.832257</td>\n",
       "      <td>590.114791</td>\n",
       "      <td>147.195839</td>\n",
       "      <td>144.355017</td>\n",
       "      <td>148.179465</td>\n",
       "      <td>148.875403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>2</td>\n",
       "      <td>14466.0</td>\n",
       "      <td>58032.0</td>\n",
       "      <td>688.0</td>\n",
       "      <td>337.0</td>\n",
       "      <td>440.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1369.498111</td>\n",
       "      <td>1276.305513</td>\n",
       "      <td>332.941295</td>\n",
       "      <td>317.56745</td>\n",
       "      <td>321.873215</td>\n",
       "      <td>292.116754</td>\n",
       "      <td>60.632767</td>\n",
       "      <td>61.876198</td>\n",
       "      <td>65.202076</td>\n",
       "      <td>60.022847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>3</td>\n",
       "      <td>41368.0</td>\n",
       "      <td>139598.0</td>\n",
       "      <td>1201.0</td>\n",
       "      <td>503.0</td>\n",
       "      <td>888.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1338.091947</td>\n",
       "      <td>1299.373271</td>\n",
       "      <td>432.829034</td>\n",
       "      <td>398.306003</td>\n",
       "      <td>401.091835</td>\n",
       "      <td>358.84984</td>\n",
       "      <td>74.837374</td>\n",
       "      <td>71.033793</td>\n",
       "      <td>80.523205</td>\n",
       "      <td>80.845266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>4</td>\n",
       "      <td>14564.0</td>\n",
       "      <td>52624.0</td>\n",
       "      <td>377.0</td>\n",
       "      <td>470.0</td>\n",
       "      <td>169.0</td>\n",
       "      <td>217.0</td>\n",
       "      <td>...</td>\n",
       "      <td>899.439956</td>\n",
       "      <td>874.837386</td>\n",
       "      <td>211.898029</td>\n",
       "      <td>189.348918</td>\n",
       "      <td>186.3333</td>\n",
       "      <td>188.292692</td>\n",
       "      <td>113.059608</td>\n",
       "      <td>113.194846</td>\n",
       "      <td>110.997393</td>\n",
       "      <td>109.83439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>5</td>\n",
       "      <td>27417.0</td>\n",
       "      <td>84084.0</td>\n",
       "      <td>760.0</td>\n",
       "      <td>550.0</td>\n",
       "      <td>487.0</td>\n",
       "      <td>242.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1231.630414</td>\n",
       "      <td>1218.998954</td>\n",
       "      <td>306.13973</td>\n",
       "      <td>295.581509</td>\n",
       "      <td>310.469726</td>\n",
       "      <td>287.78839</td>\n",
       "      <td>496.084704</td>\n",
       "      <td>502.046808</td>\n",
       "      <td>490.259298</td>\n",
       "      <td>491.171009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1198 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Metadata_ImageNumber Image_Metadata_Plate Image_Metadata_Well  \\\n",
       "0                    1                  001                  C6   \n",
       "1                    1                  001                  C6   \n",
       "2                    1                  001                  C6   \n",
       "3                    1                  001                  C6   \n",
       "4                    1                  001                  C6   \n",
       "\n",
       "  Cytoplasm_Number_Object_Number Cytoplasm_AreaShape_Area  \\\n",
       "0                              1                  32065.0   \n",
       "1                              2                  14466.0   \n",
       "2                              3                  41368.0   \n",
       "3                              4                  14564.0   \n",
       "4                              5                  27417.0   \n",
       "\n",
       "  Cytoplasm_AreaShape_BoundingBoxArea  \\\n",
       "0                             95944.0   \n",
       "1                             58032.0   \n",
       "2                            139598.0   \n",
       "3                             52624.0   \n",
       "4                             84084.0   \n",
       "\n",
       "  Cytoplasm_AreaShape_BoundingBoxMaximum_X  \\\n",
       "0                                   1018.0   \n",
       "1                                    688.0   \n",
       "2                                   1201.0   \n",
       "3                                    377.0   \n",
       "4                                    760.0   \n",
       "\n",
       "  Cytoplasm_AreaShape_BoundingBoxMaximum_Y  \\\n",
       "0                                    382.0   \n",
       "1                                    337.0   \n",
       "2                                    503.0   \n",
       "3                                    470.0   \n",
       "4                                    550.0   \n",
       "\n",
       "  Cytoplasm_AreaShape_BoundingBoxMinimum_X  \\\n",
       "0                                    750.0   \n",
       "1                                    440.0   \n",
       "2                                    888.0   \n",
       "3                                    169.0   \n",
       "4                                    487.0   \n",
       "\n",
       "  Cytoplasm_AreaShape_BoundingBoxMinimum_Y  ...  \\\n",
       "0                                     24.0  ...   \n",
       "1                                    103.0  ...   \n",
       "2                                     57.0  ...   \n",
       "3                                    217.0  ...   \n",
       "4                                    242.0  ...   \n",
       "\n",
       "  Nuclei_Texture_Variance_DAPI_3_02_256 Nuclei_Texture_Variance_DAPI_3_03_256  \\\n",
       "0                           1393.410857                           1331.583275   \n",
       "1                           1369.498111                           1276.305513   \n",
       "2                           1338.091947                           1299.373271   \n",
       "3                            899.439956                            874.837386   \n",
       "4                           1231.630414                           1218.998954   \n",
       "\n",
       "  Nuclei_Texture_Variance_GFP_3_00_256 Nuclei_Texture_Variance_GFP_3_01_256  \\\n",
       "0                           653.826838                           618.063979   \n",
       "1                           332.941295                            317.56745   \n",
       "2                           432.829034                           398.306003   \n",
       "3                           211.898029                           189.348918   \n",
       "4                            306.13973                           295.581509   \n",
       "\n",
       "  Nuclei_Texture_Variance_GFP_3_02_256 Nuclei_Texture_Variance_GFP_3_03_256  \\\n",
       "0                           606.832257                           590.114791   \n",
       "1                           321.873215                           292.116754   \n",
       "2                           401.091835                            358.84984   \n",
       "3                             186.3333                           188.292692   \n",
       "4                           310.469726                            287.78839   \n",
       "\n",
       "  Nuclei_Texture_Variance_RFP_3_00_256 Nuclei_Texture_Variance_RFP_3_01_256  \\\n",
       "0                           147.195839                           144.355017   \n",
       "1                            60.632767                            61.876198   \n",
       "2                            74.837374                            71.033793   \n",
       "3                           113.059608                           113.194846   \n",
       "4                           496.084704                           502.046808   \n",
       "\n",
       "  Nuclei_Texture_Variance_RFP_3_02_256 Nuclei_Texture_Variance_RFP_3_03_256  \n",
       "0                           148.179465                           148.875403  \n",
       "1                            65.202076                            60.022847  \n",
       "2                            80.523205                            80.845266  \n",
       "3                           110.997393                            109.83439  \n",
       "4                           490.259298                           491.171009  \n",
       "\n",
       "[5 rows x 1198 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# setting up sqlite address\n",
    "sqlite_address = f\"sqlite:///{str(plate_data)}\"\n",
    "\n",
    "# loading single cell morphology data into pycyotminer's SingleCells Object\n",
    "single_cell_profile = SingleCells(\n",
    "    sql_file=sqlite_address,\n",
    "    compartments=compartments,\n",
    "    compartment_linking_cols=linking_cols,\n",
    "    image_table_name=\"Per_Image\",\n",
    "    strata=strata,\n",
    "    merge_cols=[\"ImageNumber\"],\n",
    "    image_cols=\"ImageNumber\",\n",
    "    load_image_data=True,\n",
    ")\n",
    "\n",
    "# mering all sqlite table into a single tabular dataset (csv)\n",
    "sc_profile = single_cell_profile.merge_single_cells(\n",
    "    sc_output_files=sc_profile_path, compression_options=\"gzip\"\n",
    ")\n",
    "\n",
    "# saving single-cell morphology dataset\n",
    "sc_profile.to_csv(sc_profile_path, compression=\"gzip\")\n",
    "\n",
    "# displaying dataset\n",
    "sc_profile.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have created our merged single-cell profile, let's move on to the next step: loading our `platemaps`. \n",
    "\n",
    "Platemaps provide us with additional information that is crucial for our analysis. They contain details such as well positions, genotypes, gene names, perturbation types, and more. In other words, platemaps serve as a valuable source of supplementary information for our single-cell morphology profile."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WellRow</th>\n",
       "      <th>WellCol</th>\n",
       "      <th>well_position</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>genotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>C6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>7</td>\n",
       "      <td>C7</td>\n",
       "      <td>NF1</td>\n",
       "      <td>Null</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D</td>\n",
       "      <td>6</td>\n",
       "      <td>D6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D</td>\n",
       "      <td>7</td>\n",
       "      <td>D7</td>\n",
       "      <td>NF1</td>\n",
       "      <td>Null</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>6</td>\n",
       "      <td>E6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>E</td>\n",
       "      <td>7</td>\n",
       "      <td>E7</td>\n",
       "      <td>NF1</td>\n",
       "      <td>Null</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>F</td>\n",
       "      <td>6</td>\n",
       "      <td>F6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>F</td>\n",
       "      <td>7</td>\n",
       "      <td>F7</td>\n",
       "      <td>NF1</td>\n",
       "      <td>Null</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  WellRow  WellCol well_position gene_name genotype\n",
       "0       C        6            C6       NF1       WT\n",
       "1       C        7            C7       NF1     Null\n",
       "2       D        6            D6       NF1       WT\n",
       "3       D        7            D7       NF1     Null\n",
       "4       E        6            E6       NF1       WT\n",
       "5       E        7            E7       NF1     Null\n",
       "6       F        6            F6       NF1       WT\n",
       "7       F        7            F7       NF1     Null"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# loading plate map and display it\n",
    "platemap_df = pd.read_csv(plate_map)\n",
    "platemap_df.head(8)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotation\n",
    "\n",
    "In this step of the walkthrough, we will combine the metadata with the merged single-cell morphology dataset. To accomplish this, we will utilize the `annotation` function provided by `pycytominer`.\n",
    "\n",
    "The `annotation` function takes two inputs: the merged single-cell morphology dataset and its associated plate map. By combining these two datasets, we will generate an annotated_profile that contains enriched information.\n",
    "\n",
    "More information about the `annotation` function can be found [here](https://pycytominer.readthedocs.io/en/latest/pycytominer.html#module-pycytominer.annotate)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metadata_WellRow</th>\n",
       "      <th>Metadata_WellCol</th>\n",
       "      <th>Metadata_gene_name</th>\n",
       "      <th>Metadata_genotype</th>\n",
       "      <th>Metadata_ImageNumber</th>\n",
       "      <th>Metadata_Plate</th>\n",
       "      <th>Metadata_Well</th>\n",
       "      <th>Metadata_Cytoplasm_Parent_Cells</th>\n",
       "      <th>Metadata_Cytoplasm_Parent_Nuclei</th>\n",
       "      <th>Metadata_Cells_Number_Object_Number</th>\n",
       "      <th>...</th>\n",
       "      <th>Nuclei_Texture_Variance_DAPI_3_02_256</th>\n",
       "      <th>Nuclei_Texture_Variance_DAPI_3_03_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_00_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_01_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_02_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_03_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_00_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_01_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_02_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_03_256</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1393.410857</td>\n",
       "      <td>1331.583275</td>\n",
       "      <td>653.826838</td>\n",
       "      <td>618.063979</td>\n",
       "      <td>606.832257</td>\n",
       "      <td>590.114791</td>\n",
       "      <td>147.195839</td>\n",
       "      <td>144.355017</td>\n",
       "      <td>148.179465</td>\n",
       "      <td>148.875403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1369.498111</td>\n",
       "      <td>1276.305513</td>\n",
       "      <td>332.941295</td>\n",
       "      <td>317.56745</td>\n",
       "      <td>321.873215</td>\n",
       "      <td>292.116754</td>\n",
       "      <td>60.632767</td>\n",
       "      <td>61.876198</td>\n",
       "      <td>65.202076</td>\n",
       "      <td>60.022847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>1338.091947</td>\n",
       "      <td>1299.373271</td>\n",
       "      <td>432.829034</td>\n",
       "      <td>398.306003</td>\n",
       "      <td>401.091835</td>\n",
       "      <td>358.84984</td>\n",
       "      <td>74.837374</td>\n",
       "      <td>71.033793</td>\n",
       "      <td>80.523205</td>\n",
       "      <td>80.845266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>899.439956</td>\n",
       "      <td>874.837386</td>\n",
       "      <td>211.898029</td>\n",
       "      <td>189.348918</td>\n",
       "      <td>186.3333</td>\n",
       "      <td>188.292692</td>\n",
       "      <td>113.059608</td>\n",
       "      <td>113.194846</td>\n",
       "      <td>110.997393</td>\n",
       "      <td>109.83439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1231.630414</td>\n",
       "      <td>1218.998954</td>\n",
       "      <td>306.13973</td>\n",
       "      <td>295.581509</td>\n",
       "      <td>310.469726</td>\n",
       "      <td>287.78839</td>\n",
       "      <td>496.084704</td>\n",
       "      <td>502.046808</td>\n",
       "      <td>490.259298</td>\n",
       "      <td>491.171009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Metadata_WellRow  Metadata_WellCol Metadata_gene_name Metadata_genotype  \\\n",
       "0                C                 6                NF1                WT   \n",
       "1                C                 6                NF1                WT   \n",
       "2                C                 6                NF1                WT   \n",
       "3                C                 6                NF1                WT   \n",
       "4                C                 6                NF1                WT   \n",
       "\n",
       "  Metadata_ImageNumber Metadata_Plate Metadata_Well  \\\n",
       "0                    1            001            C6   \n",
       "1                    1            001            C6   \n",
       "2                    1            001            C6   \n",
       "3                    1            001            C6   \n",
       "4                    1            001            C6   \n",
       "\n",
       "  Metadata_Cytoplasm_Parent_Cells Metadata_Cytoplasm_Parent_Nuclei  \\\n",
       "0                               1                                3   \n",
       "1                               2                                4   \n",
       "2                               3                                5   \n",
       "3                               4                                7   \n",
       "4                               5                                8   \n",
       "\n",
       "  Metadata_Cells_Number_Object_Number  ...  \\\n",
       "0                                   1  ...   \n",
       "1                                   2  ...   \n",
       "2                                   3  ...   \n",
       "3                                   4  ...   \n",
       "4                                   5  ...   \n",
       "\n",
       "  Nuclei_Texture_Variance_DAPI_3_02_256 Nuclei_Texture_Variance_DAPI_3_03_256  \\\n",
       "0                           1393.410857                           1331.583275   \n",
       "1                           1369.498111                           1276.305513   \n",
       "2                           1338.091947                           1299.373271   \n",
       "3                            899.439956                            874.837386   \n",
       "4                           1231.630414                           1218.998954   \n",
       "\n",
       "  Nuclei_Texture_Variance_GFP_3_00_256 Nuclei_Texture_Variance_GFP_3_01_256  \\\n",
       "0                           653.826838                           618.063979   \n",
       "1                           332.941295                            317.56745   \n",
       "2                           432.829034                           398.306003   \n",
       "3                           211.898029                           189.348918   \n",
       "4                            306.13973                           295.581509   \n",
       "\n",
       "  Nuclei_Texture_Variance_GFP_3_02_256 Nuclei_Texture_Variance_GFP_3_03_256  \\\n",
       "0                           606.832257                           590.114791   \n",
       "1                           321.873215                           292.116754   \n",
       "2                           401.091835                            358.84984   \n",
       "3                             186.3333                           188.292692   \n",
       "4                           310.469726                            287.78839   \n",
       "\n",
       "  Nuclei_Texture_Variance_RFP_3_00_256 Nuclei_Texture_Variance_RFP_3_01_256  \\\n",
       "0                           147.195839                           144.355017   \n",
       "1                            60.632767                            61.876198   \n",
       "2                            74.837374                            71.033793   \n",
       "3                           113.059608                           113.194846   \n",
       "4                           496.084704                           502.046808   \n",
       "\n",
       "  Nuclei_Texture_Variance_RFP_3_02_256 Nuclei_Texture_Variance_RFP_3_03_256  \n",
       "0                           148.179465                           148.875403  \n",
       "1                            65.202076                            60.022847  \n",
       "2                            80.523205                            80.845266  \n",
       "3                           110.997393                            109.83439  \n",
       "4                           490.259298                           491.171009  \n",
       "\n",
       "[5 rows x 1202 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# annotating merged single-cell profile with metadata\n",
    "annotated_df = annotate(\n",
    "    profiles=sc_profile,\n",
    "    platemap=platemap_df,\n",
    "    join_on=[\"Metadata_well_position\", \"Image_Metadata_Well\"],\n",
    ")\n",
    "\n",
    "# save annotated profile\n",
    "annotated_df.to_csv(anno_profile_path, compression=\"gzip\")\n",
    "\n",
    "# displaying annotated profile\n",
    "annotated_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Noramlization Step\n",
    "\n",
    "The next step is to normalize our dataset using the `normalize` function provided by `pycytominer`.\n",
    "More information regards `pycytominer`'s `normalize` function can be found [here](https://pycytominer.readthedocs.io/en/latest/pycytominer.html#module-pycytominer.normalize)\n",
    "\n",
    "Normalization is a critical preprocessing step that improves the quality of our dataset. It addresses two key challenges: mitigating the impact of outliers and handling variations in value scales. By normalizing the data, we ensure that our downstream analysis is not heavily influenced by these factors.\n",
    "\n",
    "Additionally, normalization plays a crucial role in determining feature importance (which is crucial for our last step). By bringing all features to a comparable scale, it enables the identification of important features without biases caused by outliers or widely-scaled values.\n",
    "\n",
    "To normalize our annotated single-cell morphology profile, we will utilize the normalize function from pycytominer. This function is specifically designed to handle the normalization process for cytometry data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metadata_WellRow</th>\n",
       "      <th>Metadata_WellCol</th>\n",
       "      <th>Metadata_gene_name</th>\n",
       "      <th>Metadata_genotype</th>\n",
       "      <th>Metadata_ImageNumber</th>\n",
       "      <th>Metadata_Plate</th>\n",
       "      <th>Metadata_Well</th>\n",
       "      <th>Metadata_Cytoplasm_Parent_Cells</th>\n",
       "      <th>Metadata_Cytoplasm_Parent_Nuclei</th>\n",
       "      <th>Metadata_Cells_Number_Object_Number</th>\n",
       "      <th>...</th>\n",
       "      <th>Nuclei_Texture_Variance_DAPI_3_02_256</th>\n",
       "      <th>Nuclei_Texture_Variance_DAPI_3_03_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_00_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_01_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_02_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_03_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_00_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_01_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_02_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_03_256</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.277354</td>\n",
       "      <td>-0.278153</td>\n",
       "      <td>0.437881</td>\n",
       "      <td>0.406082</td>\n",
       "      <td>0.358158</td>\n",
       "      <td>0.358081</td>\n",
       "      <td>0.760504</td>\n",
       "      <td>0.736324</td>\n",
       "      <td>0.765755</td>\n",
       "      <td>0.781857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.308203</td>\n",
       "      <td>-0.353348</td>\n",
       "      <td>-0.069536</td>\n",
       "      <td>-0.073583</td>\n",
       "      <td>-0.088878</td>\n",
       "      <td>-0.116121</td>\n",
       "      <td>0.001109</td>\n",
       "      <td>0.015249</td>\n",
       "      <td>0.039691</td>\n",
       "      <td>-0.000841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.348720</td>\n",
       "      <td>-0.321969</td>\n",
       "      <td>0.088417</td>\n",
       "      <td>0.055295</td>\n",
       "      <td>0.035398</td>\n",
       "      <td>-0.009929</td>\n",
       "      <td>0.125722</td>\n",
       "      <td>0.095310</td>\n",
       "      <td>0.173753</td>\n",
       "      <td>0.182583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.914620</td>\n",
       "      <td>-0.899471</td>\n",
       "      <td>-0.260942</td>\n",
       "      <td>-0.278251</td>\n",
       "      <td>-0.301509</td>\n",
       "      <td>-0.281335</td>\n",
       "      <td>0.461035</td>\n",
       "      <td>0.463905</td>\n",
       "      <td>0.440407</td>\n",
       "      <td>0.437947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.486065</td>\n",
       "      <td>-0.431303</td>\n",
       "      <td>-0.111917</td>\n",
       "      <td>-0.108678</td>\n",
       "      <td>-0.106767</td>\n",
       "      <td>-0.123008</td>\n",
       "      <td>3.821213</td>\n",
       "      <td>3.863460</td>\n",
       "      <td>3.759005</td>\n",
       "      <td>3.797123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Metadata_WellRow  Metadata_WellCol Metadata_gene_name Metadata_genotype  \\\n",
       "0                C                 6                NF1                WT   \n",
       "1                C                 6                NF1                WT   \n",
       "2                C                 6                NF1                WT   \n",
       "3                C                 6                NF1                WT   \n",
       "4                C                 6                NF1                WT   \n",
       "\n",
       "  Metadata_ImageNumber Metadata_Plate Metadata_Well  \\\n",
       "0                    1            001            C6   \n",
       "1                    1            001            C6   \n",
       "2                    1            001            C6   \n",
       "3                    1            001            C6   \n",
       "4                    1            001            C6   \n",
       "\n",
       "  Metadata_Cytoplasm_Parent_Cells Metadata_Cytoplasm_Parent_Nuclei  \\\n",
       "0                               1                                3   \n",
       "1                               2                                4   \n",
       "2                               3                                5   \n",
       "3                               4                                7   \n",
       "4                               5                                8   \n",
       "\n",
       "  Metadata_Cells_Number_Object_Number  ...  \\\n",
       "0                                   1  ...   \n",
       "1                                   2  ...   \n",
       "2                                   3  ...   \n",
       "3                                   4  ...   \n",
       "4                                   5  ...   \n",
       "\n",
       "  Nuclei_Texture_Variance_DAPI_3_02_256  \\\n",
       "0                             -0.277354   \n",
       "1                             -0.308203   \n",
       "2                             -0.348720   \n",
       "3                             -0.914620   \n",
       "4                             -0.486065   \n",
       "\n",
       "   Nuclei_Texture_Variance_DAPI_3_03_256  \\\n",
       "0                              -0.278153   \n",
       "1                              -0.353348   \n",
       "2                              -0.321969   \n",
       "3                              -0.899471   \n",
       "4                              -0.431303   \n",
       "\n",
       "   Nuclei_Texture_Variance_GFP_3_00_256  Nuclei_Texture_Variance_GFP_3_01_256  \\\n",
       "0                              0.437881                              0.406082   \n",
       "1                             -0.069536                             -0.073583   \n",
       "2                              0.088417                              0.055295   \n",
       "3                             -0.260942                             -0.278251   \n",
       "4                             -0.111917                             -0.108678   \n",
       "\n",
       "   Nuclei_Texture_Variance_GFP_3_02_256  Nuclei_Texture_Variance_GFP_3_03_256  \\\n",
       "0                              0.358158                              0.358081   \n",
       "1                             -0.088878                             -0.116121   \n",
       "2                              0.035398                             -0.009929   \n",
       "3                             -0.301509                             -0.281335   \n",
       "4                             -0.106767                             -0.123008   \n",
       "\n",
       "   Nuclei_Texture_Variance_RFP_3_00_256  Nuclei_Texture_Variance_RFP_3_01_256  \\\n",
       "0                              0.760504                              0.736324   \n",
       "1                              0.001109                              0.015249   \n",
       "2                              0.125722                              0.095310   \n",
       "3                              0.461035                              0.463905   \n",
       "4                              3.821213                              3.863460   \n",
       "\n",
       "   Nuclei_Texture_Variance_RFP_3_02_256  Nuclei_Texture_Variance_RFP_3_03_256  \n",
       "0                              0.765755                              0.781857  \n",
       "1                              0.039691                             -0.000841  \n",
       "2                              0.173753                              0.182583  \n",
       "3                              0.440407                              0.437947  \n",
       "4                              3.759005                              3.797123  \n",
       "\n",
       "[5 rows x 1202 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# normalize dataset\n",
    "normalized_df = normalize(annotated_df)\n",
    "\n",
    "# save normalized dataset \n",
    "normalized_df.to_csv(norm_profile_path, compression=\"gzip\")\n",
    "\n",
    "# display normalized dataset\n",
    "normalized_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feature Selection\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the final section of our walkthrough, we will utilize the normalized dataset to extract important morphological features and generate a selected features profile.\n",
    "\n",
    "To accomplish this, we will make use of the `feature_select` function provided by `pycytominer`. \n",
    "Using `pycytominer`'s `feature_select` function to our dataset, we can identify the most informative morphological features that contribute significantly to the variations observed in our data. These selected features will be utilized to create our feature profile.\n",
    "\n",
    "For more detailed information about the feature_select function, its parameters, and its capabilities, please refer to the documentation available [here](https://pycytominer.readthedocs.io/en/latest/pycytominer.html#module-pycytominer.feature_select)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metadata_WellRow</th>\n",
       "      <th>Metadata_WellCol</th>\n",
       "      <th>Metadata_gene_name</th>\n",
       "      <th>Metadata_genotype</th>\n",
       "      <th>Metadata_ImageNumber</th>\n",
       "      <th>Metadata_Plate</th>\n",
       "      <th>Metadata_Well</th>\n",
       "      <th>Metadata_Cytoplasm_Parent_Cells</th>\n",
       "      <th>Metadata_Cytoplasm_Parent_Nuclei</th>\n",
       "      <th>Metadata_Cells_Number_Object_Number</th>\n",
       "      <th>...</th>\n",
       "      <th>Nuclei_Texture_Variance_DAPI_3_02_256</th>\n",
       "      <th>Nuclei_Texture_Variance_DAPI_3_03_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_00_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_01_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_02_256</th>\n",
       "      <th>Nuclei_Texture_Variance_GFP_3_03_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_00_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_01_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_02_256</th>\n",
       "      <th>Nuclei_Texture_Variance_RFP_3_03_256</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.277354</td>\n",
       "      <td>-0.278153</td>\n",
       "      <td>0.437881</td>\n",
       "      <td>0.406082</td>\n",
       "      <td>0.358158</td>\n",
       "      <td>0.358081</td>\n",
       "      <td>0.760504</td>\n",
       "      <td>0.736324</td>\n",
       "      <td>0.765755</td>\n",
       "      <td>0.781857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.308203</td>\n",
       "      <td>-0.353348</td>\n",
       "      <td>-0.069536</td>\n",
       "      <td>-0.073583</td>\n",
       "      <td>-0.088878</td>\n",
       "      <td>-0.116121</td>\n",
       "      <td>0.001109</td>\n",
       "      <td>0.015249</td>\n",
       "      <td>0.039691</td>\n",
       "      <td>-0.000841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.348720</td>\n",
       "      <td>-0.321969</td>\n",
       "      <td>0.088417</td>\n",
       "      <td>0.055295</td>\n",
       "      <td>0.035398</td>\n",
       "      <td>-0.009929</td>\n",
       "      <td>0.125722</td>\n",
       "      <td>0.095310</td>\n",
       "      <td>0.173753</td>\n",
       "      <td>0.182583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.914620</td>\n",
       "      <td>-0.899471</td>\n",
       "      <td>-0.260942</td>\n",
       "      <td>-0.278251</td>\n",
       "      <td>-0.301509</td>\n",
       "      <td>-0.281335</td>\n",
       "      <td>0.461035</td>\n",
       "      <td>0.463905</td>\n",
       "      <td>0.440407</td>\n",
       "      <td>0.437947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>NF1</td>\n",
       "      <td>WT</td>\n",
       "      <td>1</td>\n",
       "      <td>001</td>\n",
       "      <td>C6</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.486065</td>\n",
       "      <td>-0.431303</td>\n",
       "      <td>-0.111917</td>\n",
       "      <td>-0.108678</td>\n",
       "      <td>-0.106767</td>\n",
       "      <td>-0.123008</td>\n",
       "      <td>3.821213</td>\n",
       "      <td>3.863460</td>\n",
       "      <td>3.759005</td>\n",
       "      <td>3.797123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1032 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Metadata_WellRow  Metadata_WellCol Metadata_gene_name Metadata_genotype  \\\n",
       "0                C                 6                NF1                WT   \n",
       "1                C                 6                NF1                WT   \n",
       "2                C                 6                NF1                WT   \n",
       "3                C                 6                NF1                WT   \n",
       "4                C                 6                NF1                WT   \n",
       "\n",
       "  Metadata_ImageNumber Metadata_Plate Metadata_Well  \\\n",
       "0                    1            001            C6   \n",
       "1                    1            001            C6   \n",
       "2                    1            001            C6   \n",
       "3                    1            001            C6   \n",
       "4                    1            001            C6   \n",
       "\n",
       "  Metadata_Cytoplasm_Parent_Cells Metadata_Cytoplasm_Parent_Nuclei  \\\n",
       "0                               1                                3   \n",
       "1                               2                                4   \n",
       "2                               3                                5   \n",
       "3                               4                                7   \n",
       "4                               5                                8   \n",
       "\n",
       "  Metadata_Cells_Number_Object_Number  ...  \\\n",
       "0                                   1  ...   \n",
       "1                                   2  ...   \n",
       "2                                   3  ...   \n",
       "3                                   4  ...   \n",
       "4                                   5  ...   \n",
       "\n",
       "  Nuclei_Texture_Variance_DAPI_3_02_256  \\\n",
       "0                             -0.277354   \n",
       "1                             -0.308203   \n",
       "2                             -0.348720   \n",
       "3                             -0.914620   \n",
       "4                             -0.486065   \n",
       "\n",
       "   Nuclei_Texture_Variance_DAPI_3_03_256  \\\n",
       "0                              -0.278153   \n",
       "1                              -0.353348   \n",
       "2                              -0.321969   \n",
       "3                              -0.899471   \n",
       "4                              -0.431303   \n",
       "\n",
       "   Nuclei_Texture_Variance_GFP_3_00_256  Nuclei_Texture_Variance_GFP_3_01_256  \\\n",
       "0                              0.437881                              0.406082   \n",
       "1                             -0.069536                             -0.073583   \n",
       "2                              0.088417                              0.055295   \n",
       "3                             -0.260942                             -0.278251   \n",
       "4                             -0.111917                             -0.108678   \n",
       "\n",
       "   Nuclei_Texture_Variance_GFP_3_02_256  Nuclei_Texture_Variance_GFP_3_03_256  \\\n",
       "0                              0.358158                              0.358081   \n",
       "1                             -0.088878                             -0.116121   \n",
       "2                              0.035398                             -0.009929   \n",
       "3                             -0.301509                             -0.281335   \n",
       "4                             -0.106767                             -0.123008   \n",
       "\n",
       "   Nuclei_Texture_Variance_RFP_3_00_256  Nuclei_Texture_Variance_RFP_3_01_256  \\\n",
       "0                              0.760504                              0.736324   \n",
       "1                              0.001109                              0.015249   \n",
       "2                              0.125722                              0.095310   \n",
       "3                              0.461035                              0.463905   \n",
       "4                              3.821213                              3.863460   \n",
       "\n",
       "   Nuclei_Texture_Variance_RFP_3_02_256  Nuclei_Texture_Variance_RFP_3_03_256  \n",
       "0                              0.765755                              0.781857  \n",
       "1                              0.039691                             -0.000841  \n",
       "2                              0.173753                              0.182583  \n",
       "3                              0.440407                              0.437947  \n",
       "4                              3.759005                              3.797123  \n",
       "\n",
       "[5 rows x 1032 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# creating selected features profile \n",
    "features_df = feature_select(profiles=normalized_df)\n",
    "\n",
    "# saving selected features profile \n",
    "features_df.to_csv(feat_profile_path)\n",
    "\n",
    "# display selected features \n",
    "features_df.head()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Congratulations! You have successfully completed our walkthrough. We hope that this tutorial has provided you with a basic understanding of how to analyze cell morphology features using pycytominer.\n",
    "\n",
    "By following the steps outlined in this walkthrough, you have gained valuable insights into processing high-dimensional single-cell morphology data with ease using pycytominer. However, please keep in mind that pycytominer offers a wide range of functionalities beyond what we covered here. We encourage you to explore the documentation to discover more advanced features and techniques.\n",
    "\n",
    "If you have any questions or need further assistance, don't hesitate to visit the pycytominer repository and post your question in the issues section. The community is there to support you and provide guidance.\n",
    "\n",
    "Now that you have the knowledge and tools to analyze cell morphology features, have fun exploring and mining your data!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pycytominer",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
